### PR TITLE
[4.3] Fixing Itemid lookup in tags router in multilang

### PR DESCRIPTION
**Pull request by Hackwar**
Pull Request for Issue #40454.

### Summary of Changes
When running in multilanguage mode, the router falsely only checked for the current language or the default language, not for all (`*`) language when looking up the right Itemid. This executes the logic twice, once for the current language, once for the `*` language. It exits for the first matched menu item.

### Testing Instructions
1. Install Joomla
2. Install at least one additional language
3. Install multilanguage sample data
4. Install testing sample data
5. Go to for example the popular tags module. You can find it in the "Popular tags" menu item.
6. Check the URLs

### Actual result BEFORE applying this Pull Request
URLs start with /component/tags/...

### Expected result AFTER applying this Pull Request
URLs have different menu items. None start with /component/tags

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [X] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [X] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by richard67** [comment]
Does this PR conflict somehow with PR #40455 ? Or would they work together?

**Comment by Hackwar** [comment]
This PR does not conflict with #40455

**Comment by sdwjoomla** [comment]
I have tested this item :white_check_mark: successfully on 73cc2ca2744b1d0cff8fbecbfb2accc24b1b9707<br /><br />Test site with 3 languages installed (sample data) and pagination for one of the tags

With published menu item /component/ not present
Without published menu item /component/ not present

in either case no itemid<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40474">issues.joomla.org/tracker/joomla-cms/40474</a>.</sub>

**Comment by sdwjoomla** [comment]
I have tested this item :white_check_mark: successfully on 18d024d126cbc48b767547940c64c6c958ce5c6f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40474">issues.joomla.org/tracker/joomla-cms/40474</a>.</sub>

**Comment by MacJoom** [comment]
I have tested this item :red_circle: unsuccessfully on 18d024d126cbc48b767547940c64c6c958ce5c6f<br /><br />The links on the Homepage below in Popular Tags (after index.php)
Joomla 4.2.9 before Tagged-content Menu Language en-GB
/en/component/tags/tag/millions-en-gb /en/component/tags/tag/worldwide-en-gb /en/component/tags/tag/love-en-gb /en/component/tags/tag/joomla-4-en-gb

with Tagged content Language Menu en-GB with Tags 'Millions' 'Love', tagged-content within 'Millions' and 'Love' (tagged in Menu)
/en/tagged-content/millions-en-gb /en/component/tags/tag/worldwide-en-gb /en/tagged-content/love-en-gb /en/component/tags/tag/joomla-4-en-gb

after Joomla 4.3.0 - component and itemid
/en/component/tags/tag/millions-en-gb?Itemid=104 /en/component/tags/tag/worldwide-en-gb?Itemid=104 /en/component/tags/tag/love-en-gb?Itemid=104 /en/component/tags/tag/joomla-4-en-gb?Itemid=104

after applying Patch 40474 - itemid gone but still with component
/en/component/tags/tag/millions-en-gb /en/component/tags/tag/worldwide-en-gb /en/component/tags/tag/love-en-gb  /en/component/tags/tag/joomla-4-en-gb
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40474">issues.joomla.org/tracker/joomla-cms/40474</a>.</sub>

**Comment by richard67** [comment]
@Hackwar Shouldn't the `$language` bew changed to `$lang` in lines 131 and 138 as those come below the for loop? https://github.com/joomla/joomla-cms/blob/18d024d126cbc48b767547940c64c6c958ce5c6f/components/com_tags/src/Service/Router.php#L131-L138

**Comment by Hackwar** [comment]
Indeed, that needs to be changed.

**Comment by Hackwar** [comment]
done

**Comment by obuisard** [comment]
Martin @MacJoom can you double-check if that fixes it?

**Comment by obuisard** [comment]
Will this PR render #40460 obsolete?

**Comment by obuisard** [comment]
> > Will this PR render #40460 obsolete?
> 
> They both produce identical result, however I think Hannes know Joomla Routing better than me, so I have closed mine.

Fedir @Fedik, thank you so much for looking into it and providing a solution, this is really appreciated.

**Comment by MacJoom** [comment]
There is unfortunatly exactly the same result as before - beeing  on f2240e51e3 - i can switch to the 4.2.9 branch with the same db and i have the original result.

**Comment by MacJoom** [comment]
Btw. The testing instructions above are missing the step to setup a menu item with the tagged items in en-gb between steps 5 and 6 (on my test site the menu item has the alias 'tagged-content' as the links for 4.2.9 show)

**Comment by joomdonation** [comment]
To be more clear, the issue here is that if you create a menu item to link to **Tagged Items** menu item type and select two tags **A** and **B** :

- In Joomla 4.2 and earlier, link to **tag A** and **tag B** will use the above menu item as active menu item
- In Joomla 4.3, link to **tag A** and **tag B** does not use the above menu item as active menu item although the menu item contains these two tags.

There was changed to the logic of code to find active menu item for a tag for some reasons. Unless the tagged items menu item type only select a single tag, it won't be used for link to a tag like before.

**Comment by obuisard** [comment]
Doing testing and getting same results as Martin @MacJoom.

**Comment by Hackwar** [comment]
Indeed, the lookup was changed, since the old code was unable to find menu items with more than one tag assigned to them. So if you created a URL with 2 tags, it couldn't find a menu item at all. Fixing that now results in menu items with multiple tags not matching to URLs with a single tag. I personally think that tags are a filter on the content and those filters compound, so a URL with two tags should only display the content which has both tags assigned to it, but apparently I'm wrong there, since the tagging code doesn't behave that way later on in the model.

I will change the code to match it again.

**Comment by Hackwar** [comment]
Done.

**Comment by obuisard** [comment]
I have tested this item :white_check_mark: successfully on 6e3e4c3996e9e8ed1c7bdbe354d6167ce59f3cfc<br /><br />The last change seems to do it for me:

Tags that are associated to menu items are properly opening in that menu item (use of the alias of the menu item),
Tags that are not associated to menu items open in the home page (/component/tags/tag in the path).

Behavior is consistent to 4.2, no longer dimmed as b/c as far as I can tell.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40474">issues.joomla.org/tracker/joomla-cms/40474</a>.</sub>

**Comment by obuisard** [comment]
Hello @tramber91, could you take a shot at this PR and see if it addresses your issues? Merci beaucoup!

**Comment by tramber91** [comment]
@obuisard 
I have tested the file (copy raw) unfortunatly same result on my side
https://biomecanique.en-toutes-lettres.fr/fr/carte/cartographie-labo-formation/81-laboratoires-de-biomecanique/382-auctus-inria-ims
Not working, same as previous file

**Comment by richard67** [comment]
There is indeed something still wrong.

I have created a site with 2 languages, "en-GB" and "de-DE", and have installed multilingual sample data and then for each of the languages blog sample data (this is possible when switching administrator language so you can install blog sample data for each language.

Then I have created a menu item of type "Tagged Items" in the Main Menu (de-DE). That menu item has language "de-DE" and has 2 tags "Millions (de-DE)" and "Love (de-DE)" in its tag field.

I've done that on a site created with the 4.2-dev branch and on another one with the branch of this PR.

When comparing both sites, I notice following differences in the "Popular Tags" module shown at the bottom, which shows tags for both languages:

- On the 4.2-dev site, the sites which are in the "Tagged Items" menu item's tag field have an URL according to the scheme `alias-of-the-tagged-items-menu-item/alias-of-the-particular-tag`, and all other tags have URL with the `component/tags/tag/alias-of-the-particular-tag` scheme. This is the case on all pages of the site, regardless if home page or tagged items page or any other page, and regardless if the current page is de-DE or en-GB.
- On the site with 4.3-dev plus this PR, the sites which are in the "Tagged Items" menu item's tag field have an URL according to the scheme `alias-of-the-tagged-items-menu-item` when I am on a page with language "de-DE", which is the same language as the "tagged items" menu item has and the 2 tags have.
In this case, the `/alias of the particular tag` is missing at the end.
All other tags have URLs with the `component/tags/tag/alias-of-the-particular-tag` scheme.
- But on pages in language "en-GB" this is not the case, i.e. there all tags in the popular tags module have the component URLs.
- And when I am on the page which belongs to the "Tagged Items" menu item, I get for ALL tags in the module URLs which have the `alias-of-the-tagged-items-menu-item`. The tags which are in the tagged items menu item have nothing else after that part of the URL, and the other tags which are not in that menu item have the tag ID behind that, e.g. "https://www.joomla-43-dev.vmubu01.vmnet2.local/index.php/de/test-tagged-items-menu-item/2".

So we still have 3 issues (or differences compared to 4.2-dev):
 1. `/alias of the particular tag` is missing at the end of a tag URL which is in the "Tagged Items" menu item's tag field.
 2. URLs of these tags depend on the language of the current page.
 3. On the page of the "Tagged Items" menu item, the URLS for all tags in the popular tags menu are completely different to how it is on other pages.

**Comment by Hackwar** [comment]
to 1.: As I said in chat, this is actually a bugfix from my perspective, because both with and without that tag alias at the end, you get the same page, thus having duplicate content. At least with this change we aren't creating different URLs for the same page anymore.

to 3.: I'm assuming very much that this is because of #40455.

**Comment by Hackwar** [comment]
to 2.: Are you sure that the menu items for the tag views are identical in both languages?

**Comment by richard67** [comment]
@Hackwar P.S.: I don't wanna argue about what is right and what is wrong because I don't really know what's right. I only wanted to document my findings compared to 4.2. That's why I have only posted a comment but not a negative test result.

**Comment by Hackwar** [comment]
> > to 1.: As I said in chat, this is actually a bugfix from my perspective, because both with and without that tag alias at the end, you get the same page, thus having duplicate content. At least with this change we aren't creating different URLs for the same page anymore.
> 
> That makes sense of course, but I can not ready anything about this change in the description of the PR, and not all people who had the issue and want to test are in our (internal) Mattermost channel.

The question simply is, if this is a bug or not. Due to time constraints, I was not able to communicate everything extensively on every platform.

> > to 2.: Are you sure that the menu items for the tag views are identical in both languages?
> 
> @Hackwar Why should they? As said, I have created only one menu item of type "Tagged items" in language "de-DE". I have not created any in language "en-GB".

In that case, the URLs are supposed to be different. They are the URLs to the tags in different languages, so they are of course different. If you don't have a menu item for the tags, as in your case, you get the /component URLs.

**Comment by richard67** [comment]
> > > to 2.: Are you sure that the menu items for the tag views are identical in both languages?
> > 
> > 
> > @Hackwar Why should they? As said, I have created only one menu item of type "Tagged items" in language "de-DE". I have not created any in language "en-GB".
> 
> In that case, the URLs are supposed to be different. They are the URLs to the tags in different languages, so they are of course different. If you don't have a menu item for the tags, as in your case, you get the /component URLs.

It would be good if it was like this. But as I wrote, when I am on the page of that tagges items menu item, then ALL tags have an URL WITHOUT component.

The ones in the tagged menu item have `index.php/de/test-tagged-items-menu-item`, which is the URL of the current page and which would be right according to your answer to point 1.

But all other tags, i.e. those not in that menu item, they have URL's like `index.php/de/test-tagged-items-menu-item/1` or `index.php/de/test-tagged-items-menu-item/2`, and that seems not right to me.

**Comment by Hackwar** [comment]
No, the content is different, because it is in different languages. That is not something people claim is duplicate content. (I don't really want to discuss the whole idea of duplicate content, since that is not something that actually exists anymore, etc. Lets just simply try to stick to the rule "One content item only has one URL")

In any case, I did find a bug which created the single-ID URLs you described above. I will fix that now.

**Comment by Hackwar** [comment]
As far as I can see, all issues Richard reported here today are either fixed by my above commit or are unrelated to the router.

**Comment by Hackwar** [comment]
@tramber91 Can you please explain what you expect the URLs to look like, how your setup is (which menu items do you have for tags?) and are you on Joomla 4.3.0 right now, with this file being the only additional change? I can't find a report from you elsewhere and with the given information I can't help any further.

**Comment by richard67** [comment]
> So we still have 3 issues (or differences compared to 4.2-dev):
> 
>     1. `/alias of the particular tag` is missing at the end of a tag URL which is in the "Tagged Items" menu item's tag field.
> 
>     2. URLs of these tags depend on the language of the current page.
> 
>     3. On the page of the "Tagged Items" menu item, the URLS for all tags in the popular tags menu are completely different to how it is on other pages.

My issue number 3 is solved with the last commit. However, the URLs of the tags in the popular tags module are still different to "normal" pages when I am on the page of the tagged items menu item.

Number 1 and 2 are desired behaviour according to @Hackwar . I'm still not sure with number 2. The URLs in the popular tags module are different for the same tag depending on the language of the current page.

I have to stop for today. Will continue with testing tomorrow or Friday, if still necessary.

**Comment by Fedik** [comment]
@Hackwar this was an original comment about the issue https://github.com/joomla/joomla-cms/pull/40448#issuecomment-1518158231

**Comment by Hackwar** [comment]
@tramber91 can you make sure that you have the current file in this PR in your site? It is very strange that you have the current behavior.

**Comment by Hackwar** [comment]
@tramber91 would it be possible to get access to your site (or a copy) to debug this? You can reach me for example via the Joomla Mattermost or via hannes.papenberg@community.joomla.org

**Comment by tramber91** [comment]
@Hackwar sent via mail

**Comment by Hackwar** [comment]
Thank you, I found the issue. I'm looking how we can fix this in a performant way.

**Comment by Hackwar** [comment]
The problem was, that the tags view also allows to filter by parent_id, the lookup table however didn't contain any information about this or the child tags. This change now retrieves all child tags to the parent_id tag and assigns the menu item of the parent_id to its children. This can be problemativ when a tags menu item has a child tag which is also linked via tags view with a parent_id. In that case it is a matter of which menu item is processed first and that order will in the end not be guaranteed to be stable.

This is a general issue of the tags system right now. The system is half done to support nested tags and half to be a flat list. Tags for example can't have their parents alias as part of the URL and more than one tag per URL means that we have query parameters. There are a lot of inconsistencies in the system and for a proper, stable routing, we would need something like the CategoryNode class for tags. Rewriting the whole tag system would really be awesome...

In any case, please test and then hopefully we can get this issue solved.

File: components/com_tags/src/Service/Router.php [hunk 1]
```diff
@@ -271,13 +291,48 @@ protected function buildLookup()
                 $id = $item->query['id'];
                 sort($id);
                 $this->lookup[$item->language]['tag'][implode(',', $id)] = $item->id;
+
+                foreach ($id as $i) {
+                    $this->lookup[$item->language]['tag'][$i] = $item->id;
```
Comment (Fedik) [inline components/com_tags/src/Service/Router.php:294]: I think we need to be careful with this. I have tried the same but reverted.
The reason, when Menu would have 2 following menu item:
```
/tag1(Itemid X) => Contain only Tag1
/tag1-and-tag2 (Itemid Y) => Contain Tag1 and Tag2
```
Then lookup for Tag1 will hold only 1 itemid (Itemid Y) because it is last in the order.

Maybe it would be better to store these items in separated array? Kind of:
```php
$this->lookup[$item->language]['tag.partial'][$i] = $item->id;
```

And then in `process()` check first `['tag']` and then `['tag.partial']` lookup array.

Comment (Hackwar) [reply components/com_tags/src/Service/Router.php:294]: The tags view only has a single ID, the tag view does support more than one and will provide the same content regardless if you provide the single tag ID for a multiple-tag-menuitem or not. That is what I'm talking about the whole time, that the tag view displays all content items which have at least one of the selected tags tagged to it, instead of showing only content items which contain all tags.

Long story short: This is B/C the way it is, even though that it is also broken. But without a real major release which allows us to break B/C, we wont get that solved.

**Comment by MacJoom** [comment]
Thank you very much @Hackwar - can you give me instructions how to setup my test site? with my current setup i still get the same url for both tags, which is probably ok - but not the same as in 4.2.9. i would love to see how to setup the menu with the tagged items like Betrand did.

**Comment by richard67** [comment]
> with my current setup i still get the same url for both tags, which is probably ok - but not the same as in 4.2.9.

@MacJoom As far as I understood @Hackwar that's intended for the case that you have both tags in the same tagged items menu item because when you click that link you come to that page in both cases. In 4.2.9 there was a `/name-of-the-tag` part appended to the URL, which now with this PR will be removed. But as said, when clicking that link on 4.2.9 you come to the page of the tagged menu item, and when doing that on the 4.3-dev with this PR applied, it goes to the same page, they just have different URLs but show information for both tags.

That makes sense to me and is not really a b/c break since the old 4.2.9 and the new URL from this PR do the same, and the PR solves the duplicate content you have with the old 4.2.9 URL's.

So if that is the only difference between the URLs from this PR and from 4.2.9, that would be ok to me.

Any other differences should be discussed further.

Anyway, your idea with instructions on how to reproduce that special scenario is good. Let's see what comes.

**Comment by richard67** [comment]
@Hackwar There is one unexpected difference in the URLs in the Popular Tags module between 4.2.9 and your PR in addition to the expected one mentioned in my previous comment:

When I am at the home page or any other page except of the page which belongs to the "Tagged Items" menu item,, everything is as expected.

But when I am on the page of the menu item of type "Tagged Items", I get different results:
- On 4.2.9 everything is like on any other page of the site.
- On 4.3-dev + this PR, the URL of all tags in that module have an URL on the "Tagged Items" page. The ones which are in that menu item have the right URL, but those which are not have the URL of the tagged menu items page plus `/name-of-the-tag` appended, e.g. `index.php/en/test-tagged-items-menu-item-en-gb/millions-de-de`.

When I comment our or remove the following code, this difference is fixed, and the tags which are not in the tagged items menu item have the URL with component in it, like it is on all other pages:

https://github.com/joomla/joomla-cms/blob/1c29079a0ed1f0fa9b749e438374ffb74e99d5bb/components/com_tags/src/Service/Router.php#L142-L148

So either this code should be removed, or something is wrong in that `if` condition which needs to be fixed.

Could you check that, and maybe you or @tramber91 could check if it still works for his scenario when that part of the code is removed?

Thanks in advance, and thanks a lot for all your efforts up to now.

File: components/com_tags/src/Service/Router.php [hunk 2]
```diff
@@ -89,47 +89,67 @@ public function preprocess($query)
...
-                if (!is_array($query['id'])) {
-                    $query['id'] = [$query['id']];
-                }
-
-                $id = ArrayHelper::toInteger($query['id']);
-                sort($id);
-
-                if (isset($this->lookup[$language]['tag'][implode(',', $id)])) {
-                    $query['Itemid'] = $this->lookup[$language]['tag'][implode(',', $id)];
+        foreach ([$lang, '*'] as $language) {
```
Comment (richard67) [inline components/com_tags/src/Service/Router.php]:
```suggestion
        foreach (array_unique([$lang, '*']) as $language) {
```
If multilanguage is not enabled, `$lang`will be `'*'`, so the `foreach` will loop over `['*', '*']`. We can avoid this with using array_unique.

File: components/com_tags/src/Service/Router.php [hunk 3]
```diff
@@ -89,47 +89,67 @@ public function preprocess($query)
...
+                    sort($id);
+
+                    if (isset($this->lookup[$language]['tag'][implode(',', $id)])) {
+                        $query['Itemid'] = $this->lookup[$language]['tag'][implode(',', $id)];
+                        break;
+                    } else {
+                        foreach ($id as $i) {
+                            if (isset($this->lookup[$language]['tag'][$i])) {
+                                $query['Itemid'] = $this->lookup[$language]['tag'][$i];
+                                break;
```
Comment (richard67) [inline components/com_tags/src/Service/Router.php]:
```suggestion
                                break 2;
```
I think we can break the outer loop over the languages here, too. But I'm not 100 % sure. I did not really notice any change when testing this suggestion. But if it works it is better for performance.

**Comment by Hackwar** [comment]
> can you give me instructions how to setup my test site?

Create one tag and then another tag with the first tag as parent. Tag a content item with the child tag. Create a `tags` menu item and set the parent tag to your parent tag. Now you have the setup from Bertrand.

Regarding the URL: Yes, it is expected, that these are the same right now. It's due to tags being simply broken.

Regarding the code mentioned by @richard67: Yes, that needs to be removed.

**Comment by richard67** [comment]
@tramber91 @MacJoom Could you test again with the latest changes? Thanks in advance.

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 796a8c03b7173a70868fd36bf0e532c3d48fa900<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40474">issues.joomla.org/tracker/joomla-cms/40474</a>.</sub>

**Comment by tramber91** [comment]
ok for me too, same and good result. New file seems less heavy!!

**Comment by richard67** [comment]
> ok for me too, same and good result. New file seems less heavy!!

@tramber91 Thanks for checking and reporting back. On thing you should know: As this pull request here will very likely be too late for the upcoming 4.0.1, it will go into 4.0.2. This means that when you will update your site to 4.0.1, the fix will get lost. You can safely put again the file from this PR on your site to fix it. When you later update to 4.0.2 when that is released, the fix should be included in that version, so you won’t need to apply the fix again.

@MacJoom Could you also test again with respect to the one expected difference we explained above, and also test other scenarios like monolingual site? I have successfully tested that, too, but the more tests the better.

**Comment by tramber91** [comment]
Thanks @richard67 
i understand the manip
Suppose you speak about 4.3.1 and 4.3.2

**Comment by richard67** [comment]
> Thanks @richard67 i understand the manip Suppose you speak about 4.3.1 and 4.3.2

@tramber91 Yes, 4.3.1 and 4.3.2.

**Comment by obuisard** [comment]
Hi @tramber91, would you mind reporting your test results in the issue tracker (https://issues.joomla.org/tracker/joomla-cms/40474)? That way, with 2 successful tests and all the work that has been done in this PR by Hannes and all the tests you all have done, we can merge it and test it more widely until the 4.3.2 release. Thank you!

**Comment by tramber91** [comment]
@obuisard, last file checked with Joomla 4.3.2
It is ok, same result as 4.3.1
I keep it in my live updated website.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40474">issues.joomla.org/tracker/joomla-cms/40474</a>.</sub>

**Comment by obuisard** [comment]
> @obuisard, last file checked with Joomla 4.3.2 It is ok, same result as 4.3.1 I keep it in my live updated website.
> 
> This comment was created with the [J!Tracker Application](https://github.com/joomla/jissues) at [issues.joomla.org/tracker/joomla-cms/40474](https://issues.joomla.org/tracker/joomla-cms/40474).

Does that mean it's successful for you @tramber91? Just want to make sure. And if it is successful, report it with the 'Test this' button on the issue tracker. Thank you :-)

**Comment by tramber91** [comment]
I have tested this item :white_check_mark: successfully on 796a8c03b7173a70868fd36bf0e532c3d48fa900<br /><br />tested too with success<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40474">issues.joomla.org/tracker/joomla-cms/40474</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40474">issues.joomla.org/tracker/joomla-cms/40474</a>.</sub>

**Comment by joeforjoomla** [comment]
I have tested this item :red_circle: unsuccessfully on 796a8c03b7173a70868fd36bf0e532c3d48fa900<br /><br />All tag links broken, each tag has the same link as the menu item itself.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40474">issues.joomla.org/tracker/joomla-cms/40474</a>.</sub>

**Comment by Hackwar** [comment]
@joeforjoomla If you create a tag menu item and select all the tags in the menu item, that is expected behavior. Before the changes you had the same page, just with all the different URLs.

**Comment by richard67** [comment]
@joeforjoomla I can confirm what @Hackwar write in his previous comment. And just by the way: A broken link is a link which results in a 404. Do you get a 404? I am pretty sure: no. So why do you write about broken links?

**Comment by Hackwar** [comment]
@joeforjoomla Can you please answer @richard67 and me? This is a rather important issue and we are all waiting for you to either substantiate your claim or to merge this PR to be ready in time for 4.3.2. If you don't respond in 3 days, I will suggest to ignore your report.

**Comment by joeforjoomla** [comment]
@Hackwar @richard67 i would expect to have tag links as before:

http://joomla4/en/menualias/tag/tag1.html
http://joomla4/en/menualias/tag/tag2.html

and not

http://joomla4/en/menualias.html

Why this? How to have a tag link then?

**Comment by Hackwar** [comment]
I reviewed the whole code of tags again and you are right @joeforjoomla. I'm going to see what I can do here.

**Comment by joeforjoomla** [comment]
> @joeforjoomla Please check if this change fixes your issues.

@Hackwar Just perfect!

**Comment by joeforjoomla** [comment]
I have tested this item :white_check_mark: successfully on d5a6863aca7f90f7e754161f5e2f411d5792f64c<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40474">issues.joomla.org/tracker/joomla-cms/40474</a>.</sub>

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on d5a6863aca7f90f7e754161f5e2f411d5792f64c<br /><br />URLs now after the last change are like they were in 4.2.9, i.e. everything as expected now.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40474">issues.joomla.org/tracker/joomla-cms/40474</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40474">issues.joomla.org/tracker/joomla-cms/40474</a>.</sub>

**Comment by obuisard** [comment]
Thank you Hannes @Hackwar and all testers, your help is greatly appreciated. Great team work.
